### ISSUE TITLE
shindo gem: hung test and eventual OS crash

### ISSUE THREAD

**Issue by gogainda** [oracle/truffleruby#2175]
when you run shindo [gem](https://github.com/geemus/shindo) test suit it freezes on "execution" test and after 10 if you try to cancel it (^C) it crashes the MacOS

It's ok to crush but not OS
@geemus 

Only happens on Truffleruby

**Comment by eregon** [comment]
@bjfish Could you try to reproduce?
It sounds very weird that the OS would crash.
No process should be able to crash the OS, of course, so it might be partly a macOS bug.

**Comment by eregon** [comment]
I could reproduce it on Linux, it seems running that test suite creates many subprocess, much like a fork bomb.

**Comment by eregon** [comment]
The issue is that the detection for which tests to run behaves differently on TruffleRuby, and ends up running all tests for every subprocess, instead of just the passed file. And since the tests create subprocesses running `shindo`, it recursively runs the entire test suite until out of processes or memory.

This is because of a `||=` assignment within `for`, which ends up on TruffleRuby creating a `tests` local variable within the `for`'s frame and not a local variable for the top-level scope of the file.
https://github.com/geemus/shindo/blob/master/lib/shindo/bin.rb#L30-L33

Adding `tests = nil` before the `for` loop works around it, and then all tests pass.

Minimal reproducer:
```
$ ruby -e 'for i in [1]; a||=1; end; p a'
1 # MRI
nil # TruffleRuby
```
(`ruby -e 'for i in [1]; a=1; end; p a'` is `1` for both so that case is already handled correctly)

**Comment by geemus** [comment]
Hmm. Thanks for digging in to this. If you have something working, could you do a PR with it? I wanted to make sure I'm not missing anything in what you describe as the fix. Thanks!

**Comment by eregon** [comment]
@geemus It's literally just
```diff
diff --git a/lib/shindo/bin.rb b/lib/shindo/bin.rb
index acf19a3..2bceb59 100644
--- a/lib/shindo/bin.rb
+++ b/lib/shindo/bin.rb
@@ -21,6 +21,7 @@ end
 
 helpers = Dir.glob(File.join('tests', '**', '*helper.rb')).sort_by {|helper| helper.count(File::SEPARATOR)}
 tags = []
+tests = nil
 for argument in ARGV
   if argument.match(/^[\+\-]/)
     tags << argument
```

I'd make a PR, but there seems to be no license and I'm rather busy with other things.
Of course, we should fix this in TruffleRuby, but it seems a good idea to initialize the variable more clearly too.

**Comment by geemus** [comment]
Thanks for clarifying. I thought that was what you meant, but it wasn't obvious to me how/why that would solve the problem. That change certainly seems fine/safe, so I'll try to do that soon.

Apologies about the license, it should be MIT (as I have done with other things), but it must have gotten missed/forgotten somewhere along the line. I'll try to fix that soon.

**Comment by geemus** [comment]
* License/contrib info added here: https://github.com/geemus/shindo/commit/4d9989f39b7b5f0120063064fec3f3bbe34309f1
* `tests = nil` added here: https://github.com/geemus/shindo/commit/e628f0395a15410081d1800dc309d5a80515521e

If someone can confirm that does the expected behavior (I don't have truffleruby setup to do it myself), I can cut a release.

Thanks!

**Comment by bjfish** [comment]
@geemus I ran the test suite successfully on TruffleRuby.

**Comment by gogainda** [comment]
adding truffleruby support here https://github.com/geemus/shindo/pull/19

**Comment by geemus** [comment]
@bjfish Thanks for confirming.

@gogainda Thanks for adding that, in hindsight that seems like a pretty obvious way for me to have tested it as well, but it totally slipped my mind.

Fix released in v0.3.9.
